File: DNS_Resolver.c
/*
Montana Pawek
Resources used:
    Example Files in Assignment 5 Folder:
        lookup.c
    Shared Files in OneDrive:
        Pthreads folder:
            hello.c
            hello_arg2.c
            join.c
        pthreadshared folder:
            goodcount.c
            bettercount.c
            bestcount.c
        IPC folder:
            shm_mutex.c
    Edstem Files:
        producerconsumer_many.c
    Man Pages:
        pthreads
        pthread_create
        pthread_join
        pthread_mutex_init
        pthread_mutex_lock
        usleep
        random
        valgrind
        pthread_cond_t
        pthread_cond_wait
        pthread_cond_init
        pthread_cond_signal

Skeleton of code based on provided lookup.c file
* File: lookup.c
* Author: Andy Sayler
* Project: CSCI 3753 Programming Assignment 2
* Create Date: 2012/02/01
* Modify Date: 2012/02/01
* Description:
* 	This file contains the reference non-threaded
*      solution to this assignment.
*  
*/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <pthread.h>
#include <errno.h>
#include <unistd.h>
#include <sys/time.h>
 
#include "util.h"
#include "multi-lookup.h"
 
#define MINARGS 3
#define USAGE "<inputFilePath> <outputFilePath>"
#define INPUTFS "%1024s"

// Function called by first pthread_create; takes strings from input files and loads them into the buffer
void *requester (void *shared_v)
{
    // Recast variable back to struct from void *
    struct shared_variables *sv = (struct shared_variables *) shared_v;

    // Variable used to hold opening of input file error messages
    char errorstr[MAX_NAME_LENGTH];

    // String to hold hostname
    char hostname[MAX_NAME_LENGTH];
    
    // Loop Through Input Files
    for (int i = 1; i < (sv->argc - 1); i++)
    {        
        // Error Check: Open Input File
        // Borrowed from lookup.c
        // Set struct input file pointer to current input file
        sv->inputfp = fopen(sv->argv[i], "r");

        // If input file won't open:
        if(!sv->inputfp){
            sprintf(errorstr, "Error Opening Input File: %s", sv->argv[i]);
            perror(errorstr);
            break;
        }	

        // First critical section; when we put input into bounded buffer
        // While there are more lines to read into the hostname string:
        while (fscanf (sv->inputfp, INPUTFS, hostname) > 0)
        {
            // Try to get lock for the buffer
            pthread_mutex_lock (&sv->buffer);

            // Check to make sure buffer is not full; if it is, wait until it isn't
            while (sv->count >= MAX_INPUT_FILES)
            {
                // Wait until signal is received buffer has space for more names
                // This conditional wait was taken from Assignment 6 code
                pthread_cond_wait (&sv->not_full, &sv->buffer);
            }

            // Copy strings into buffer, use head pointer as it's the first free spot
            strcpy (sv->shared_buffer [sv->head], hostname);
            // Set head pointer to the next available spot; if it's outside the buffer wraparound to the beginning
            sv->head = (++(sv->head) % MAX_INPUT_FILES);

            // Iterate count by 1 to reflect buffer now has one more address
            sv->count++;

            // Signal pthread_wait that buffer has more names to pull
            pthread_cond_signal (&sv->not_empty);

            // After copying to the buffer, unlock
            pthread_mutex_unlock (&sv->buffer);
        }
       
        // Close Input File
        fclose (sv->inputfp);
    }

    // Set requester flag to 1, indicating the requester has finished all it's work and resolver can exit once it's done
    // Critical section; make sure to lock/unlock when changing struct values that other processes will use
    // Note that we use broadcast here to signal all waiting resolver threads instead of just 1
    // Errors would occur where not all resolvers would get the signal and exit
    pthread_mutex_lock (&sv->buffer);
    sv->requesterDone = 1;
    pthread_cond_broadcast (&sv->not_empty);
    pthread_mutex_unlock (&sv->buffer);

    // Exit thread if we reach the end
    pthread_exit (NULL);
}

// Function called by second pthread_create; takes strings from buffer and checks if they're legit. If they are, puts them in results
void *resolver (void * shared_v)
{
    // String to hold buffer string
    char lookupName[MAX_NAME_LENGTH];

    // String to hold IP string
    char firstipstr[INET6_ADDRSTRLEN];

    // Recast variable back to struct from void *
    struct shared_variables *sv = (struct shared_variables *) shared_v;

    // Infinitely loop until requester signals it's done
    while (1)
    {
        // Need to lock the buffer to read from it and remove a string
        pthread_mutex_lock (&sv->buffer);

        // While the buffer is empty but the requester isn't done yet, wait
        while (sv->count < 1 && !sv->requesterDone) 
        {
            // Wait to be signalled the buffer has strings
            // Code taken from Assignment 6
            pthread_cond_wait(&sv->not_empty, &sv->buffer);
        }

        // Once out of the wait loop, check to see if the requester has finished and the buffer is empty. If it has and if it is, exit
        if (sv->count < 1 && sv->requesterDone) 
        {
            pthread_mutex_unlock (&sv->buffer);
            pthread_exit (NULL);
        }

        // Take string from buffer, targeting oldest existing name at the tail, and copy into local variable
        strcpy (lookupName, sv->shared_buffer[sv->tail]);
        // Mark the spot we removed the string from the buffer as ready to be filled
        sv->shared_buffer[sv->tail][0] = '\0';  
        // Set tail pointer to the next available spot; if it's outside the buffer wraparound to the beginning
        sv->tail = (++(sv->tail) % MAX_INPUT_FILES);
                
        // Decrement count to reflect one less address in buffer
        sv->count--;

        // Signal that the buffer has open spaces that can be filled
        pthread_cond_signal (&sv->not_full);

        // Done checking the buffer, unlock it
        pthread_mutex_unlock (&sv->buffer);
        
        // Lookup code borrowed from lookup.c
        // Lookup the hostname and get IP string
        if(dnslookup (lookupName, firstipstr, sizeof(firstipstr)))
        {
            // Error Check: Lookup error
            fprintf (stderr, "dnslookup error: %s\n", lookupName);
            fflush (sv->outputfp);
            strncpy (firstipstr, "", sizeof(firstipstr));
        }
        

        // Now we need to write results to the results file, so we attempt to acquire the lock
        pthread_mutex_lock (&sv->results);

        // Write to Output File, flush to make sure it happens immediately
        fprintf (sv->outputfp, "%s,%s\n", lookupName, firstipstr);
        fflush (sv->outputfp);

        // After writing, unlock
        pthread_mutex_unlock (&sv->results);
    }

    // If thread somehow reaches here (it shouldn't ever), exit
    pthread_exit(NULL);
}
 
int main(int argc, char* argv[])
{
    // Initialize struct
    struct shared_variables sv;

    // Initialize shared_buffer first chars to zero so code works
    for (int q = 0; q < 10; q++)
    {
        sv.shared_buffer[q][0] = 0;
    }

    // Inititalize sv variables
    sv.argc = argc;
    sv.argv = argv;
    sv.count = 0;
    sv.requesterDone = 0;
    sv.head = 0;
    sv.tail = 0;

    // Initialize thread pointer variables
    // Only need one producer thread pointer, as we only need one requester
    // We use an array for the consumer thread pointers, as we will be using a minimum of two resolvers
    // return_value holds pthread_create value to check for errors
    pthread_t p_thread;
    pthread_t c_threads[MAX_RESOLVER_THREADS];
    int return_value;

    // Initialize both mutex locks, and conditional lock
    pthread_mutex_init(&sv.buffer, NULL);
    pthread_mutex_init(&sv.results, NULL);
    pthread_cond_init(&sv.not_full, NULL);
    pthread_cond_init(&sv.not_empty, NULL);

     
    // Error Check: Check Arguments 
    // Borrowed from lookup.c
    // Check to make sure we have the proper number of command line arguments
    if(argc < MINARGS)
    {
        fprintf(stderr, "Not enough arguments: %d\n", (argc - 1));
        fprintf(stderr, "Usage:\n %s %s\n", argv[0], USAGE);
        return EXIT_FAILURE;
    }
 
    // Error Check: Open Output File
    // Borrowed from lookup.c
    // Check to make sure we can open the output file to store results
    // Set struct output file pointer
    sv.outputfp = fopen(argv[(argc-1)], "w");

    // If we can't open the file:
    if(!sv.outputfp)
    {
        perror("Error Opening Output File");
        return EXIT_FAILURE;
    }

    // Output file for time results:
    FILE* time_output = fopen ("C_DNSResolver.txt", "a");
    
    // Make sure time results file actually opened:
    if (!time_output)
    {
        printf ("Error opening file");
        exit (-1);
    }

    // Calculate time taken; initialize clock and start timer
    // Code borrowed from CSCI440 github repo timing.c example
    struct timespec time_start, time_end;
    clock_gettime (CLOCK_MONOTONIC, &time_start);

    // Create Threads       
    // Check for error when making threads; pthread_create passes 0 back if threads were created successfully
    // Inspiration taken from hello.c example file
    // First argument: Pointer to created thread.
    // Second argument: Don't worry about it; NULL
    // Third Argument: Function pointer to function that runs when thread is created
    // Fourth Argument: Argument to function, (void *) t

    // Create our single producer/requester thread
    return_value = pthread_create (&p_thread, NULL, requester, (void *) &sv);

    // Thread creation error checking
    if (return_value)
    {
        fprintf(stderr, "Requester thread creation error; #%d\n", return_value);
        exit(-1);
    }

    // Create second set of threads consumer/resolver threads to read bounded buffer and try to lookup 
    for (int m = 0; m < MAX_RESOLVER_THREADS; m++)
    {
        return_value = pthread_create (&c_threads[m], NULL, resolver, (void *) &sv);

        // Thread creation error checking
        if (return_value)
        {
            fprintf(stderr, "Resolver thread creation error; #%d\n", return_value);
            exit(-1);
        }
    }
    
    // Join requester thread
    pthread_join (p_thread, NULL);

    // Join resolver threads
    for (int n = 0; n < MAX_RESOLVER_THREADS; n++)
    {
        pthread_join(c_threads[n], NULL);
    }

    // Finish timer as work is done
    clock_gettime (CLOCK_MONOTONIC, &time_end);

    // Calculate time taken
    // NOTE: kept getting negative time results, so we have to modify this part to make sure that doesn't happen
    double time_taken; // = ((end_time.tv_sec - start_time.tv_sec) + (end_time.tv_nsec - start_time.tv_nsec)) / 1e9;

    // Error occurs if the end_time.tv_nsec is less than start_time.tv_nsec due to wraparound errors; if statement checks if that's the case
    if (time_end.tv_nsec < time_start.tv_nsec) 
    {
        time_taken = ((time_end.tv_sec - time_start.tv_sec - 1) + (time_end.tv_nsec + 1e9 - time_start.tv_nsec)) / 1e9;
    } 
        
    else 
    {
        time_taken = ((time_end.tv_sec - time_start.tv_sec) + (time_end.tv_nsec - time_start.tv_nsec)) / 1e9;
    }

    // Print time taken to output file
    fprintf (time_output, "%lf\n", time_taken );
 
    // Close Output Files
    fclose (sv.outputfp);
    fclose (time_output);
 
    return EXIT_SUCCESS;
}

File: DNS_Resolver.py
# Montana Pawek
# Resources used:
# Mainly for Python Syntax, Python multiprocessing, Python DNS lookup
    # https://www.geeksforgeeks.org/multiprocessing-python-set-1/
    # https://www.geeksforgeeks.org/multiprocessing-python-set-2/
    # https://www.geeksforgeeks.org/synchronization-pooling-processes-python/
    # https://www.geeksforgeeks.org/python-program-find-ip-address/
    # https://www.digitalocean.com/community/tutorials/python-multiprocessing-example
    # https://www.digitalocean.com/community/tutorials/python-get-ip-address-from-hostname
    # https://docs.python.org/3/library/multiprocessing.html
    # https://docs.python.org/3/library/socket.html
    # https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Process
    # https://docs.python.org/3/library/multiprocessing.html#multiprocessing-programming
    # https://docs.python.org/3/library/sys.html
    # https://docs.python.org/3/tutorial/inputoutput.html
    # https://stackoverflow.com/questions/2575760/python-lookup-hostname-from-ip-with-1-second-timeout
    # https://github.com/codemistic/Data-Structures-and-Algorithms/blob/main/Python%20script%20to%20display%20ip%20address%20and%20host%20name.py
    # https://superfastpython.com/multiprocessing-mutex-lock-in-python/


import multiprocessing as mp
import socket
import sys
import time

# Constants, set to same as C file
MAX_INPUT_FILES = 10
MAX_RESOLVER_PROCESSES = 10
MIN_RESOLVER_PROCESSES = 2


# Shared Object: 
    # Manager allows shared memory between multiple processes
    # buffer is the shared bounded buffer
    # buffer_lock is the lock for the bounded buffer
    # file_lock is the lock for the output file
    # not_full and not_empty are the two conditions for the buffer lock
    # locks required due to critical sections, same as in C program, conditional locks help with requester/resolver
    # Requester_done_flag is the boolean value that is tripped when the requester thread has done all it's work
    # NOTE: Python recommends not using lock.acquire/lock.release manually, as using 'with' with locks automatically releases them when the action has been completed
    # Had to switch conditional lock/buffer to mp instead of manager to avoid errors
class shared_variables:
    def __init__ (self, manager):
        self.buffer = manager.list ()
        self.buffer_lock = manager.Lock ()
        self.file_lock = manager.Lock ()
        self.not_empty = manager.Condition (self.buffer_lock)
        self.not_full = manager.Condition (self.buffer_lock)
        self.done_flag = manager.Value ('b', False)

# NOTE: We use python's DNS lookup feature here because it's super easy to implement, and is certainly easier than figuring out how to
# get python to run util.c or translating util.c to python
def dnslookup (hostname):
    try:
        return socket.gethostbyname (hostname)
    
    # If the lookup doesn't work, mimic the C code logic where it leaves an empty string after the name where the IP address would go
    except socket.gaierror:
        return ""


def requester (sv, input_files):
    # For loop for each file in input_files array
    for file_path in input_files:

        # Try to open file
        try:
            # Open input file as read-only, python file notation suggests using 'with' keyword to make sure file closes properly if error occurs
            with open (file_path, 'r') as infile:

                # Keep reading lines from the file as long as there are lines to read
                for line in infile:

                    # Take string from line w/ whitespaces removed
                    hostname = line.strip ()

                    # Attempt to acquire the buffer_lock if it has the not_full condition (e.g. it has space to put more names in):
                    with sv.not_full:
                        # While the buffer is full, wait
                        while len (sv.buffer) >= MAX_INPUT_FILES:
                            # Unlocks the lock and waits until it's signalled that it can continue
                            sv.not_full.wait ()

                        # Once there's space in the buffer insert the string 
                        sv.buffer.append(hostname)

                        # Notify the not_empty condition there is a string in the buffer
                        sv.not_empty.notify()

        # If we can't open file, print error message:
        except Exception as e:
            print (f"Error reading {file_path}: {e}", file = sys.stderr)

    # Must obtain buffer lock to switch flag to done and indicate requester is finished once all input has been taken
    with sv.buffer_lock:
        sv.done_flag.value = True

    # Before requester exits, it notifies resolvers that they should check the buffer for more strings
    with sv.not_empty:
        sv.not_empty.notify_all()


def resolver (sv, output_file): 
    # Endlessly loop until flag is tripped
    while True:

        # Check conditional variable to make sure we have strings to remove from the buffer
        with sv.not_empty:

            # If buffer is empty and has no strings:
            while not sv.buffer:

                # Check if done_flag is true, if so, we leave the function
                if sv.done_flag.value:
                    return

                # Otherwise, wait until conditional variable signals there are strings in the buffer
                sv.not_empty.wait ()

            # Remove the first string from the buffer
            hostname = sv.buffer.pop (0)

            # After a string has been removed from the buffer, signal there is room in the buffer
            sv.not_full.notify ()

        # Call the dnslookup function on the string to determine IP address
        # Returns IP address is successful, empty string on failure
        ip = dnslookup (hostname)

        # Obtain output file lock, and write results to the output file
        with sv.file_lock:
            with open(output_file, 'a') as f:
                f.write(f"{hostname},{ip}\n")


def main ():
    # Error Check: Check Arguments
    # Make sure we have program call, input file(s), and output file
    if len (sys.argv) < 3:
        print ("Not enough arguments")
        sys.exit (1)
   
    # Initialize input and output files based on command line arguments
    # Reference info: https://docs.python.org/3/library/sys.html
    # Assign input_files all command line arguments after the program name until the last argument, not including the last argument
    # Assign output_file the last command line argument
    input_files = sys.argv[1:-1]
    output_file = sys.argv[-1]

    # Opens outputfile and clears it of previous IPs
    with open (output_file, 'w'):
        pass

    # Initialize manager so processes can share object
    manager = mp.Manager ()
    sv = shared_variables (manager)

    # Start timer
    start_time = time.time_ns ()

    # Create and start requester process, only need 1
    # Arguments: requester = function to be invoked by start/run, args = data to be passed in
    # cont...: including the shared object with the buffer info and locks, and the array of input files we grab names from
    requester_process = mp.Process (target = requester, args = (sv, input_files))
    requester_process.start ()

    # Determine number of resolver processes, and initialize array to hold them
    num_resolvers = MAX_RESOLVER_PROCESSES
    res_procs_array = []

    # Create and start resolver processes
    # For loop that iterates for total number of resolvers being used
    # Arguments: resolver = function to be invoked by start/run, args = data to be passed in 
    # cont...: including the shared object with the buffer info and locks, and the file we output the results to
    # Then, append the current resolver process address to end of the res_procs_array list
    for _ in range (num_resolvers):
        resolver_processes = mp.Process (target = resolver, args = (sv, output_file))
        resolver_processes.start ()
        res_procs_array.append (resolver_processes)

    # Wait for processes to finish, requester first, then join the resolver processes after
    requester_process.join ()
    for resolver_processes in res_procs_array:
        resolver_processes.join ()

    # End timer, calculate time taken
    end_time = time.time_ns ()
    time_taken = (end_time - start_time) / (10 ** 9)

    # Define and open output file to store results in
    with open ("PyDNSResolver.txt", "a") as output:
        # output.write (f"Time taken: {time_taken:.6f} seconds - Estimated Pi: {pi_estimate:.10f}\n")
        output.write (f"{time_taken:.6f}\n")

if __name__ == '__main__':
    # I ran into errors that suggested this code be used for Linux/WSL compatibility
    mp.set_start_method ("fork")  
    main ()